### DISCUSSION TITLE
Support for dictionaries/maps

### DISCUSSION THREAD

**Discussion by fencepost-error** [BlackZork/mqmgateway#52] [Ideas]
One thing I noticed while creating the templates (from the other discussion thread) is that an awful lot of devices... actually all of them, use magic values for parameters, typically baud rate, data formats, display values, etc.  It would be really nice to be able to use dictionaries/maps in converters to map the magic numbers to something more meaningful.  I'm principally thinking of their use in templates here so you don't need to look up all the magic values when you want to do something.  For example for baud rate the converter might be:

```
converter: map( "{1: 2400, 2: 4800, 3: 9600, 4: 19200, 5: 38400}" )
```

Then you could read/write the baud rate as 9600 rather than having to remember that it's 3 for this type of device, 2 for another type of device, and 5 for a third type of device.

**Comment by BlackZork** [comment]
You can partially workaround it using `expr.evaluate` with switch instruction. Unfortunately it works only for state section.

std.map would be useful, but to implement it I need much better argument parser than the current one, with at least support for arguments in single and double quotes. Named arguments would be nice too.

**Reply by fencepost-error** [reply]
state-only is fine, so far the templates I've written are all read-only (i.e. for the state section) as a safety measure.  Most of the devices I'm playing with *cough*Aliexpress*cough* are ones where you want to carefully monitor and control writes to holding registers with configuration data because anything done wrong could require a lot of work to reset the device to a working state, if it's resettable at all.

For `expr.evaluate` do you mean something like:

```
expr.evaluate( "switch { case R0 == 1 : 2400; case R0 == 2 : 4800; case R0 == 3 : 9600; case R0 == 4 : 19200; case R0 == 5 : 38400; default: 0 }" )
```

**Reply by BlackZork** [reply]
Yes. Does it work for you?

**Reply by fencepost-error** [reply]
Thanks!

I'm travelling for awhile and the documentation for all the magic values involved is back home so it'll be awhile before I can post the next progress report, but it'll be nice to have meaningful values displayed for a number of registers that are currently just small integers.

**Reply by fencepost-error** [reply]
Finally got back, and unfortunately it doesn't work, a test expression like:
```
converter: expr.evaluate("R0 + R1")
```
works fine but something like:
```
converter: expr.evaluate( "switch { case R0 == 1 : 2400; case R0 == 2 : 4800; default: 0; }" )
```
gives:
```
[CRITICAL] yaml-cpp: error at line 530, column 58: illegal map value
```

**Reply by BlackZork** [reply]
I think you need to escape  at least : char, maybe { } too.

**Reply by fencepost-error** [reply]
It looks like you can't escape a colon in a string with '\\' even with multiple levels of them, e.g. if '\\\\' is converted to '\\' by something.  I've managed to get it to pass by using \&#58; in place of the colon, but then I get NaN as the result which I assume is exprtk saying it doesn't like the resulting string, possibly because it's still seeing the \&#58; in there rather than a colon.

Just tried it with \u003A and that doesn't work either.

**Comment by BlackZork** [comment]
std.map added in b849a5d (upcoming 2.8.0). See [description in std converters section](https://github.com/BlackZork/mqmgateway?tab=readme-ov-file#standard-converters).

**Reply by fencepost-error** [reply]
Thanks, will play with it and report back in a day or two.

**Reply by fencepost-error** [reply]
OK, one minor request, for running bleeding-edge versions of the code (or in fact any version) having a '--version' argument for modmqttd would be nice, at the moment I can't see any way to check which version is actually running.

**Reply by fencepost-error** [reply]
Unfortunately it doesn't seem to work, using:
```
        converter: std.map( '{ 1: "cm", 2: "mm", 3: "mPa", 4: "Pa", 5: "kPa", 6: "MA" }' )
```
(tried both with and without spaces after the colon) gives:
```
[CRITICAL] config error(line 830): Key 1 is empty
```
without the space and:
```
[CRITICAL] yaml-cpp: error at line 830, column 33: illegal map value
```
with the space.  Column 33 is the `1: "cm"` portion.  [YAML lint](https://www.yamllint.com/) says it's valid YAML for the form without the space after the colon.

**Reply by BlackZork** [reply]
I fixed other problem related to string values, but found something that looks like yaml-cpp bug. You cannot put a space between key and value, `'{ 1:"cm", 2:"mm" }'` works OK.

**Reply by BlackZork** [reply]
Added more tests and fixed bug with str map value that starts with a digit (example `{1:"2"}`).

There is a way to specify a map with spaces as text in new line:

```
converter: >
  std.map( '{ 1: "cm", 2: "mm", 3: "mPa", 4: "Pa", 5: "kPa", 6: "MA" }' )
```

Also, it looks like my parser treats curly braces are optional. I consider it not as a bug, but a feature :-)

**Reply by BlackZork** [reply]
> I fixed other problem related to string values, but found something that looks like yaml-cpp bug. 

Sorry, not a yaml-cpp bug. I wrongly read that it should work with space between key and a value. It is not valid yaml, so you need to use `: >` format in this case like in my example above.

**Reply by fencepost-error** [reply]
Thanks, that's working now:

```
        converter: std.map( '{ 1:"cm", 2:"mm", 3:"mPa", 4:"Pa", 5:"kPa", 6:"MA" }' )
```

![mqtt](https://github.com/user-attachments/assets/27859aee-5ec5-46fb-8723-f4790cc75c87)

However when using a zero value in the map:

```
        converter: std.map( '{ 0:"nnnn", 1:"nnn.n", 2:"nn.nn", 3:"n.nnn" }' )
#        converter: std.map( '{ 1:"nnn.n", 2:"nn.nn", 3:"n.nnn" }' )
```

this gives:

```
[CRITICAL] config error(line 838): Register value  already mapped
```

Is there a default mapping already set up for the value 0?

**Reply by BlackZork** [reply]
Fixed in [320737](https://github.com/BlackZork/mqmgateway/commit/3207377320656b584a4968ed6d4bbfa0b4d9b93f)

**Reply by fencepost-error** [reply]
Yup, all good now, thanks!

**Reply by fencepost-error** [reply]
Thanks again for adding this, it makes things so much easier than having to remember a bunch of arbitrary numeric values for different things:

![map](https://github.com/user-attachments/assets/0f388359-6943-427f-a929-8d5b23913dc0)
